hail/python/hail/expr/functions.py

```diff
@@ -4357,6 +4357,214 @@ def approx_equal(x, y, tolerance=1e-6, absolute=False, nan_same=False):
+
+    Returns
+    -------
+    :class:`.Int32Expression` or :class:`.Int64Expression`
+    """
+    return _bit_op(x, y, '&')
+
+
+@typecheck(x=expr_oneof(expr_int32, expr_int64), y=expr_oneof(expr_int32, expr_int64))
+def bit_or(x, y):
+    """Bitwise or `x` and `y`.
+
+    Examples
+    --------
+    >>> hl.eval(hl.bit_or(5, 3))
+    7
+
+    Notes
+    -----
+    See `the Python wiki <https://wiki.python.org/moin/BitwiseOperators>`__
+    for more information about bit operators.
+
+
+    Parameters
+    ----------
+    x : :class:`.Int32Expression` or :class:`.Int64Expression`
+    y : :class:`.Int32Expression` or :class:`.Int64Expression`
+
+    Returns
+    -------
+    :class:`.Int32Expression` or :class:`.Int64Expression`
+    """
+    return _bit_op(x, y, '|')
+
+
+@typecheck(x=expr_oneof(expr_int32, expr_int64), y=expr_oneof(expr_int32, expr_int64))
+def bit_xor(x, y):
+    """Bitwise exclusive-or `x` and `y`.
+
+    Examples
+    --------
+    >>> hl.eval(hl.bit_xor(5, 3))
+    6
+
+    Notes
+    -----
+    See `the Python wiki <https://wiki.python.org/moin/BitwiseOperators>`__
+    for more information about bit operators.
+
+
+    Parameters
+    ----------
+    x : :class:`.Int32Expression` or :class:`.Int64Expression`
+    y : :class:`.Int32Expression` or :class:`.Int64Expression`
+
+    Returns
+    -------
+    :class:`.Int32Expression` or :class:`.Int64Expression`
+    """
+    return _bit_op(x, y, '^')
+
+
+@typecheck(x=expr_oneof(expr_int32, expr_int64), y=expr_oneof(expr_int32, expr_int64))
+def bit_xor(x, y):
+    """Bitwise exclusive-or `x` and `y`.
+
+    Examples
+    --------
+    >>> hl.eval(hl.bit_xor(5, 3))
+    6
+
+    Notes
+    -----
+    See `the Python wiki <https://wiki.python.org/moin/BitwiseOperators>`__
+    for more information about bit operators.
+
+
+    Parameters
+    ----------
+    x : :class:`.Int32Expression` or :class:`.Int64Expression`
+    y : :class:`.Int32Expression` or :class:`.Int64Expression`
+
+    Returns
+    -------
+    :class:`.Int32Expression` or :class:`.Int64Expression`
+    """
+    return _bit_op(x, y, '^')
+
+
+@typecheck(x=expr_oneof(expr_int32, expr_int64), y=expr_int32)
+def bit_lshift(x, y):
+    """Bitwise left-shift `x` by `y`.
+
+    Examples
+    --------
+    >>> hl.eval(hl.bit_lshift(5, 3))
+    40
+
+    >>> hl.eval(hl.bit_lshift(1, 8))
+    256
+
+    Notes
+    -----
+    See `the Python wiki <https://wiki.python.org/moin/BitwiseOperators>`__
+    for more information about bit operators.
+
+
+    Parameters
+    ----------
+    x : :class:`.Int32Expression` or :class:`.Int64Expression`
+    y : :class:`.Int32Expression` or :class:`.Int64Expression`
+
+    Returns
+    -------
+    :class:`.Int32Expression` or :class:`.Int64Expression`
+    """
+    return _bit_op(x, y, '<<', unify_type=False)
+
+
+@typecheck(x=expr_oneof(expr_int32, expr_int64), y=expr_int32)
+def bit_rshift(x, y):
+    """Bitwise right-shift `x` by `y`.
```

Comment:
Moreover, I _think_ python elides a logical right shift operator (called `>>>` in Java/Scala) because Python ints are arbitrary width, ergo, it's maybe not clear where the high zeros are showing up. We (hail), have fixed-width integers so a logical right shift operator can be sensibly defined.

Can we add `bit_logical_rshift(x, y)` as well? Or add a `logical` keyword argument? Or maybe `preserve_sign`?

Reply:
@danking Thanks for doing the work for me ;)

Reply:
added logical argument and docs